saleor/warehouse/management.py

```diff
@@ -47,32 +57,54 @@ def allocate_stock(
             "quantity_allocated__sum"
         ]
 
+    variant_to_stocks: Dict[str, List[StockData]] = defaultdict(list)
+    for stock_data in stocks.values("product_variant", "pk", "quantity"):
+        variant = stock_data.pop("product_variant")
+        variant_to_stocks[variant].append(StockData(**stock_data))
+
+    allocations: List[Allocation] = []
+    for line_info in order_lines_info:
+        stocks = variant_to_stocks[line_info.variant.pk]  # type: ignore
+        allocations.extend(
+            _create_allocations(line_info, stocks, quantity_allocation_for_stocks)
+        )
+
+    if allocations:
+        Allocation.objects.bulk_create(allocations)
+
+
+def _create_allocations(
+    line_info: "OrderLineData",
+    stocks: List[StockData],
+    quantity_allocation_for_stocks: dict,
+):
+    quantity = line_info.quantity
     quantity_allocated = 0
     allocations = []
+    for stock_data in stocks:
+        quantity_allocated_in_stock = quantity_allocation_for_stocks.get(
+            stock_data.pk, 0
+        )
 
-    for stock in stocks:
-        quantity_allocated_in_stock = quantity_allocation_for_stocks.get(stock.pk, 0)
-
-        quantity_available_in_stock = stock.quantity - quantity_allocated_in_stock
+        quantity_available_in_stock = stock_data.quantity - quantity_allocated_in_stock
 
         quantity_to_allocate = min(
             (quantity - quantity_allocated), quantity_available_in_stock
         )
         if quantity_to_allocate > 0:
             allocations.append(
                 Allocation(
-                    order_line=order_line,
-                    stock=stock,
+                    order_line=line_info.line,
+                    stock_id=stock_data.pk,
                     quantity_allocated=quantity_to_allocate,
                 )
             )
 
             quantity_allocated += quantity_to_allocate
             if quantity_allocated == quantity:
-                Allocation.objects.bulk_create(allocations)
-                break
+                return allocations
     if not quantity_allocated == quantity:
-        raise InsufficientStock(order_line.variant)
+        raise InsufficientStock(line_info.variant)
```

Comment:
I think we should raise this error for all invalid lines instead of the first-line without proper stock.